lib/Sema/CodeSynthesis.cpp

```diff
@@ -1643,6 +1643,46 @@ void synthesizeAccessorBody(AbstractFunctionDecl *fn, void *) {
   llvm_unreachable("bad synthesized function kind");
 }
 
+static void maybeAddMemberwiseDefaultArg(ParamDecl *arg, VarDecl *var,
+                    SmallVectorImpl<DefaultArgumentInitializer *> &defaultInits,
+                                         unsigned paramSize, ASTContext &ctx) {
+  // First and foremost, if this is a constant don't bother.
+  if (var->isLet())
+    return;
+
+  // We can only provide default values for patterns binding a single variable.
+  // i.e. var (a, b) = getSomeTuple() is not allowed.
+  if (!var->getParentPattern()->getSingleVar())
+    return;
+
+  // If we don't have an expression initializer or silgen can't assign a default
+  // initializer, then we can't generate a default value. An example of where
+  // silgen can assign a default is var x: Int? where the default is nil.
+  if (!var->getParentPatternBinding()->isDefaultInitializable())
+    return;
+
+  // We can add a default value now.
+
+  // Give this some bogus context right now, we'll fix it after making
+  // the constructor.
+  auto *initDC = new (ctx) DefaultArgumentInitializer(
+    arg->getDeclContext(), paramSize);
+
+  defaultInits.push_back(initDC);
+
+  // Set the default value to the variable. When we emit this in silgen
+  // we're going to call the variable's initializer expression.
+  arg->setStoredProperty(var);
+  arg->setDefaultArgumentKind(DefaultArgumentKind::StoredProperty);
+
+  // If the type is T? and has no initial value, then set the default arg kind
+  // to nil literal. This is useful when we need to print the constructor.
+  // Note, this will always be the sugared T? because we don't default init an
+  // explicit Optional<T>.
+  if (isa<OptionalType>(var->getType().getPointer()) && !var->hasInitialValue())
+    arg->setDefaultArgumentKind(DefaultArgumentKind::NilLiteral);
```

Comment:
To shed some light on this change, the way printing `StoredProperty` works is by getting the parent initializer, but for `T?` cases there is no initializer. We don't build out any `NilLiteralExpr`, but we set the arg kind so that printing will just give us `nil`.

Reply:
... in fact this already appears to correctly handle the lazy property case. Maybe just remove the `areLazyAccessorsSynthesized` check above and see what happens?

Also, can you restructure this code a bit so that setDefaultArgumentKind() is only called once? It's a bit confusing to see it overwritten later. I'm guessing you don't really need to call setStoredProperty() either in the NilLiteral case?